at-date-lib/src/main/kotlin/com/amerharb/atdate/Encode.kt

```diff
@@ -0,0 +1,366 @@
+		}
+
+		else -> Pair(providedZoneLevel ?: ZoneLevel.Level0, null)
+	}
+
+	return Moment(
+		rangeLevel = rangeLevel,
+		resolutionLevel = resolutionLevel,
+		zoneLevel = zoneLevel,
+		accuracy = accuracy,
+		leapSecondsFlag = leapSecondsFlag,
+		date = dateULong,
+		time = timeULong,
+		zone = zoneULong,
+		plusLeapSeconds = if (leapSecondsFlag == 0.toUByte()) null else pl,
+		minusLeapSeconds = if (leapSecondsFlag == 0.toUByte()) null else ml
+	)
+}
+
+fun encodePeriod(input: String): Period {
+	// takes input like "@P3D {d:1 t:0 l:0-0}@"
+	// takes input like "@P1DT05:00:00 {d:1 t:5 l:0-0}@"
+	// and returns an Period object
+	val ad = input.trim().substringAfter("@P").substringBefore("@").trim()
+	if (ad.trim() == "+tp") {
+		return Period.getPositiveTinyPeriod()
+	} else if (ad.trim() == "-tp") {
+		return Period.getNegativeTinyPeriod()
+	}
+	val dayTime = ad.substringBefore("{").trim()
+	val dayTimeArr = dayTime.split("T")
+	val dayPart = dayTimeArr[0].substringBefore("D")
+	val timePart = if (dayTimeArr.size > 1) dayTimeArr[1] else ""
+	val prop = ad.substringAfter("{").substringBefore("}")
+	val propArr = prop.trim().split(" ").map { it.trim() }
+
+	val dValue = propArr.find { it.startsWith("d:") }?.substringAfter(":")
+	val providedRangeLevel = RangeLevel.values().find { it.no == dValue?.toUByte() }
+
+	val tValue = propArr.find { it.startsWith("t:") }?.substringAfter(":")
+	val providedResolutionLevel = ResolutionLevel.values().find { it.no == tValue?.toUByte() }
+
+	val lValue = propArr.find { it.startsWith("l:") }?.substringAfter(":")
+	val lList = lValue?.split("-")?.map { it.toULong() } ?: listOf(0UL, 0UL)
+	val pl = lList[0]
+	val ml = lList[1]
+	val leapSecondsFlag: UByte = when {
+		(pl == 0UL && ml == 0UL) -> 0U
+		(pl < 0xFFUL && ml < 0xFFUL) -> 1U
+		(pl < 0xFFFFUL && ml < 0xFFFFUL) -> 2U
+		(pl < 0xFF_FFFFUL && ml < 0xFF_FFFFUL) -> 3U
+		(pl < 0xFFFF_FFFFUL && ml < 0xFFFF_FFFFUL) -> 4U
+		(pl < 0xFF_FFFF_FFFFUL && ml < 0xFF_FFFF_FFFFUL) -> 5U
+		(pl < 0xFFFF_FFFF_FFFFUL && ml < 0xFFFF_FFFF_FFFFUL) -> 6U
+		(pl < 0xFF_FFFF_FFFF_FFFFUL && ml < 0xFF_FFFF_FFFF_FFFFUL) -> 7U
+		(pl < 0xFFFF_FFFF_FFFF_FFFFUL && ml < 0xFFFF_FFFF_FFFF_FFFFUL) -> 8U
+		else -> throw Exception("leap second over 8 bytes is not supported")
+	}
+
+	val sign = !dayPart.startsWith("-")
+	val day = dayPart.toLong()
+	val rangeLevel = providedRangeLevel ?: getSuitableRangeLevelForPeriod(day)
+	val dateULong = when (rangeLevel) {
+		RangeLevel.Level0 -> 0UL
+		RangeLevel.Level1 -> {
+			// only most right 15 bits are used
+			day.absoluteValue.toULong() and 0b01111111_11111111UL
+		}
+
+		RangeLevel.Level2 -> day.absoluteValue.toULong()
+		RangeLevel.Level3 -> day.absoluteValue.toULong()
+		RangeLevel.Level4 -> day.absoluteValue.toULong()
+	}
+
+	val (resolutionLevel, timeULong) = if (timePart.trim() != "") {
+		val (hour, min, sec, secFraction, precision) = destructTimePart(timePart)
```

Comment:
use calculateTimeResolution function

Reply:
The way the current encoding work only support up to LEVEL10 in order to support higher level we need to use array of UByte instead of ULong.
thats need to be done in diffreant PR